### PR TITLE
Code format

### PR DESCRIPTION
**Pull request by QianrenLi**
ADD an outline symbol provider and a default code formatter.

## Symbol Provider
After get a document:

`document` --> `genTexElements()` --> `TeXElement[]` --> `elementsToSymbols()` --> `vscode.DocumentSymbol[]`

The `document` will first be parsed by `parse` into `Ast.node`, after that, `parseNode()` iteratively parse the node into a vector `TeXElement[]`. The `TeXElement[]` will be further propossed by `hierarchyStructFormat()` to offer a structured symbol tree.

During `genTexElements()`, three helpers `macroToStr()`,  `envToStr()` and `argContentToStr()` convert arguments of node, which parsed by unified-latex from document, to string.

## Code Formatter
Use the prettier offered by `unified-latex`

### CODE REVIEW

**Comment by QianrenLi** [comment]
I am not sure how to reduce the file size. Is that can be done by deleting the packect.lock json?

**Comment by iamhyc** [comment]
> I am not sure how to reduce the file size. Is that can be done by deleting the packect.lock json?

Can you try re-generated the lock file? The two `@unified-latex` packages introduces too many dependencies

File: src/intellisense/texDocFormatter.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        // Return the TextEdit as an array
+        return [edit];
+        });
+    }
+
+    get triggers(){
+        const latexSelector = ['latex', 'latex-expl3', 'pweave', 'jlweave', 'rsweave'].map( (id) => {
+            return { language: id };
+         });
+        const texDocFormatter = new TexDocFormatter();
```
Comment (iamhyc) [inline src/intellisense/texDocFormatter.ts:57]: why new itself? use `this` or make the function static

File: src/intellisense/texDocFormatter.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+export class TexDocFormatter implements vscode.DocumentFormattingEditProvider {
+    provideDocumentFormattingEdits(document: vscode.TextDocument, options: vscode.FormattingOptions, token: vscode.CancellationToken): vscode.ProviderResult<vscode.TextEdit[]> {
+        // Get the entire document text
+        const text = document.getText();
+
+        // Format the LaTeX code here
+        return format(text).then(formattedText => {
+        // const formattedText = format(text);
```
Comment (iamhyc) [inline src/intellisense/texDocFormatter.ts:30]: what is this comment for?

File: src/intellisense/texDocFormatter.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+/*
+    * @param {string} source - The LaTeX source code to be formatted.
+    * @param {object} options - The options for Prettier.
+    * @returns {Promise<string>} - The formatted LaTeX source code.
+*/
+// https://github.com/siefkenj/latex-parser-playground/blob/master/src/async-worker/parsing-worker.ts#L35-L43
+async function format(source = "", options = {}) {
+    return Prettier.format(source, {
+        printWidth: 80,
```
Comment (iamhyc) [inline src/intellisense/texDocFormatter.ts:15]: what is this magic number "80" for? why?

File: src/intellisense/texDocFormatter.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        useTabs: true,
```
Comment (iamhyc) [inline src/intellisense/texDocFormatter.ts:16]: tabs or space? why set `true` here?

File: src/intellisense/texDocSymbolProvider.ts [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+async function parseNode(
+        node: Ast.Node,
+        root: { children: TeXElement[] }) {
+    const attributes = {
+        lineFr: (node.position?.start.line ?? 1) - 1,
+        lineTo: (node.position?.end.line ?? 1) - 1,
+        children: []
+    };
+    let element: TeXElement | undefined;
+    if (node.type === 'macro'&& defaultStructure.includes(node.content) && node.args?.[3].openMark === '{') {
```
Comment (iamhyc) [inline src/intellisense/texDocSymbolProvider.ts:231]: use `switch-case` syntax. And could you please improve the readability?

File: src/intellisense/texDocSymbolProvider.ts
Comment (iamhyc) [inline src/intellisense/texDocSymbolProvider.ts:1]: move all function definitions before the class, especially when they are used in the class and you are not going to export them

**Review by iamhyc — CHANGES_REQUESTED** [review]

File: src/intellisense/texDocFormatter.ts (hunk 3 above)
Comment (QianrenLi) [reply src/intellisense/texDocFormatter.ts:15]: This number is the length of formated string length, like this 
image


And it is 80 in default, it is revealed for future application (if any)

File: src/intellisense/texDocSymbolProvider.ts (hunk 5 above)
Comment (QianrenLi) [reply src/intellisense/texDocSymbolProvider.ts:231]: By using switch and case, and some if-else statements are transformed into forms like `case 'section'` which might improve the readability a little bit.

**Comment by QianrenLi** [comment]
After I remove the content of packet.lock file, the regenerated file seems to be larger...
image

**Review by iamhyc — CHANGES_REQUESTED** [review]

File: src/intellisense/texDocFormatter.ts (hunk 3 above)
Comment (QianrenLi) [reply src/intellisense/texDocFormatter.ts:15]: The unit is space.

File: src/intellisense/texDocFormatter.ts [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    * @returns {Promise<string>} - The formatted LaTeX source code.
+*/
+// https://github.com/siefkenj/latex-parser-playground/blob/master/src/async-worker/parsing-worker.ts#L35-L43
+async function format(source = "", options: vscode.FormattingOptions ) {
+    let useTabs = !options?.insertSpaces;
+    let printWidth = Number(options.printWidth) || 80;
+    return Prettier.format(source, {
+        printWidth: printWidth,
+        useTabs: useTabs,
+        ...options,
```
[the remark this answers is not in the document]
Comment (QianrenLi) [reply src/intellisense/texDocFormatter.ts:21]: Works, at least the value is parsed into which can see by formatting into `printWidth = 10`, but I do not found any places in vscode to set this value.

[the remark this answers is not in the document]
Comment (iamhyc) [reply src/intellisense/texDocFormatter.ts:21]: Don't guess. You are unpacking a vscode type and pass to a third party package. How could it supposed to be work? Be serious!

**Review by iamhyc — CHANGES_REQUESTED** [review]

File: src/intellisense/texDocFormatter.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (QianrenLi) [reply src/intellisense/texDocFormatter.ts:21]: image
I mean this part which is transformed here. Of cause, the thrid party package definitely not deal with the options offered by vscode. But here, the vscode only offer two options in default, one is `insertSpaces` which is here, the other is `tabSize` which to me, I do not found a corresponding entry in the third party.

[the remark this answers is not in the document]
Comment (iamhyc) [reply src/intellisense/texDocFormatter.ts:21]: parse the tabsize, remove the printWidth

**Review by iamhyc — APPROVED** [review]

File: src/intellisense/texDocFormatter.ts [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        // Create a TextEdit to replace the selected text with the formatted text
+        const edit = new vscode.TextEdit(range, formattedText);
+        // Return the TextEdit as an array
+        return [edit];
+        });
+    }
+
+    get triggers(){
+        const latexSelector = ['latex', 'latex-expl3', 'pweave', 'jlweave', 'rsweave'].map( (id) => {
+            return { language: id };
```
Comment (iamhyc) [inline src/intellisense/texDocFormatter.ts:54]: also restrict the scheme to `ROOT_NAME`

reference: https://code.visualstudio.com/api/references/vscode-api#DocumentSelector

File: src/intellisense/texDocSymbolProvider.ts [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+            if (section.children.length > 0) {
+                symbol.children = this.elementsToSymbols(section.children);
+            }
+        });
+        return symbols;
+    }
+
+    get triggers(){
+        const latexSelector = ['latex', 'latex-expl3', 'pweave', 'jlweave', 'rsweave'].map((id) => {
+            return { language: id };
```
Comment (iamhyc) [inline src/intellisense/texDocSymbolProvider.ts:360]: also restrict the scheme to ROOT_NAME

see reference above

**Review by iamhyc — CHANGES_REQUESTED** [review]
please make sure this two features only work when the resource scheme is `ROOT_NAME`, i.e. `overleaf-workshop`

**Comment by iamhyc** [comment]
Please double check and then check again on any possible missing flaws.

Anyway, if there is any bug report issue related to this PR, I hope you could make fast response.

File: src/intellisense/texDocFormatter.ts (hunk 7 above)
Comment (QianrenLi) [reply src/intellisense/texDocFormatter.ts:54]: OK

File: src/intellisense/texDocSymbolProvider.ts (hunk 8 above)
Comment (QianrenLi) [reply src/intellisense/texDocSymbolProvider.ts:360]: OK

**Review by iamhyc — APPROVED** [review]
Again, I am not going to verify the changes. Make sure by yourself.

Besides, cleanup the typo in your commit information.

**Comment by iamhyc** [comment]
> Please do not merge until issue #51 is closed. After that fix is confirmed, merge as you like.

That fix is confirmed. Please merge this PR at your time, and we can then prepare the next release.

**Comment by iamhyc** [comment]
Please also help notify in #35 the feature is complete and close it, once `v0.5.5` is available.

**Comment by QianrenLi** [comment]
Sorry for being a little late in reading the message ...

**Pull request merged**